File: api/src/main/java/org/apache/iceberg/expressions/Evaluator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iceberg.expressions;

import java.io.Serializable;
import java.util.Comparator;
import java.util.Set;
import org.apache.iceberg.StructLike;
import org.apache.iceberg.expressions.ExpressionVisitors.BoundVisitor;
import org.apache.iceberg.types.Comparators;
import org.apache.iceberg.types.Types.StructType;
import org.apache.iceberg.util.NaNUtil;

/**
 * Evaluates an {@link Expression} for data described by a {@link StructType}.
 *
 * <p>Data rows must implement {@link StructLike} and are passed to {@link #eval(StructLike)}.
 *
 * <p>This class is thread-safe.
 */
public class Evaluator implements Serializable {
  private final Expression expr;

  public Evaluator(StructType struct, Expression unbound) {
    this.expr = Binder.bind(struct, unbound, true);
  }

  public Evaluator(StructType struct, Expression unbound, boolean caseSensitive) {
    this.expr = Binder.bind(struct, unbound, caseSensitive);
  }

  public boolean eval(StructLike data) {
    return new EvalVisitor().eval(data);
  }

  private class EvalVisitor extends BoundVisitor<Boolean> {
    private StructLike struct;

    private boolean eval(StructLike row) {
      this.struct = row;
      return ExpressionVisitors.visitEvaluator(expr, this);
    }

    @Override
    public Boolean alwaysTrue() {
      return true;
    }

    @Override
    public Boolean alwaysFalse() {
      return false;
    }

    @Override
    public Boolean not(Boolean result) {
      return !result;
    }

    @Override
    public Boolean and(Boolean leftResult, Boolean rightResult) {
      return leftResult && rightResult;
    }

    @Override
    public Boolean or(Boolean leftResult, Boolean rightResult) {
      return leftResult || rightResult;
    }

    @Override
    public <T> Boolean isNull(Bound<T> valueExpr) {
      return valueExpr.eval(struct) == null;
    }

    @Override
    public <T> Boolean notNull(Bound<T> valueExpr) {
      return valueExpr.eval(struct) != null;
    }

    @Override
    public <T> Boolean isNaN(Bound<T> valueExpr) {
      return NaNUtil.isNaN(valueExpr.eval(struct));
    }

    @Override
    public <T> Boolean notNaN(Bound<T> valueExpr) {
      return !NaNUtil.isNaN(valueExpr.eval(struct));
    }

    @Override
    public <T> Boolean lt(Bound<T> valueExpr, Literal<T> lit) {
      Comparator<T> cmp = lit.comparator();
      return cmp.compare(valueExpr.eval(struct), lit.value()) < 0;
    }

    @Override
    public <T> Boolean lt(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      Comparator<T> cmp = Comparators.forType(valueExpr.ref().type().asPrimitiveType());
      return cmp.compare(valueExpr.eval(struct), valueExpr2.eval(struct)) < 0;
    }

    @Override
    public <T> Boolean ltEq(Bound<T> valueExpr, Literal<T> lit) {
      Comparator<T> cmp = lit.comparator();
      return cmp.compare(valueExpr.eval(struct), lit.value()) <= 0;
    }

    @Override
    public <T> Boolean ltEq(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      Comparator<T> cmp = Comparators.forType(valueExpr.ref().type().asPrimitiveType());
      return cmp.compare(valueExpr.eval(struct), valueExpr2.eval(struct)) <= 0;
    }

    @Override
    public <T> Boolean gt(Bound<T> valueExpr, Literal<T> lit) {
      Comparator<T> cmp = lit.comparator();
      return cmp.compare(valueExpr.eval(struct), lit.value()) > 0;
    }

    @Override
    public <T> Boolean gt(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      Comparator<T> cmp = Comparators.forType(valueExpr.ref().type().asPrimitiveType());
      return cmp.compare(valueExpr.eval(struct), valueExpr2.eval(struct)) > 0;
    }

    @Override
    public <T> Boolean gtEq(Bound<T> valueExpr, Literal<T> lit) {
      Comparator<T> cmp = lit.comparator();
      return cmp.compare(valueExpr.eval(struct), lit.value()) >= 0;
    }

    @Override
    public <T> Boolean gtEq(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      Comparator<T> cmp = Comparators.forType(valueExpr.ref().type().asPrimitiveType());
      return cmp.compare(valueExpr.eval(struct), valueExpr2.eval(struct)) >= 0;
    }

    @Override
    public <T> Boolean eq(Bound<T> valueExpr, Literal<T> lit) {
      Comparator<T> cmp = lit.comparator();
      return cmp.compare(valueExpr.eval(struct), lit.value()) == 0;
    }

    @Override
    public <T> Boolean eq(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      Comparator<T> cmp = Comparators.forType(valueExpr.ref().type().asPrimitiveType());
      return cmp.compare(valueExpr.eval(struct), valueExpr2.eval(struct)) == 0;
    }

    private <T> void validateDataTypes(Bound<T> valueExpr, Bound<T> valueExpr2) {
      if (valueExpr.ref().type().typeId() != valueExpr2.ref().type().typeId()) {
        throw new IllegalArgumentException(
            "Cannot compare different types: "
                + valueExpr.ref().type()
                + " and "
                + valueExpr2.ref().type());
      }
    }

    @Override
    public <T> Boolean notEq(Bound<T> valueExpr, Literal<T> lit) {
      return !eq(valueExpr, lit);
    }

    @Override
    public <T> Boolean notEq(Bound<T> valueExpr, Bound<T> valueExpr2) {
      validateDataTypes(valueExpr, valueExpr2);
      return !eq(valueExpr, valueExpr2);
    }

    @Override
    public <T> Boolean in(Bound<T> valueExpr, Set<T> literalSet) {
      return literalSet.contains(valueExpr.eval(struct));
    }

    @Override
    public <T> Boolean notIn(Bound<T> valueExpr, Set<T> literalSet) {
      return !in(valueExpr, literalSet);
    }

    @Override
    public <T> Boolean startsWith(Bound<T> valueExpr, Literal<T> lit) {
      T evalRes = valueExpr.eval(struct);
      return evalRes != null && ((String) evalRes).startsWith((String) lit.value());
    }

    @Override
    public <T> Boolean notStartsWith(Bound<T> valueExpr, Literal<T> lit) {
      return !startsWith(valueExpr, lit);
    }
  }
}

File: python/iceberg/api/expressions/evaluator.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import threading

from .binder import Binder
from .expressions import ExpressionVisitors


class Evaluator(object):

    def __init__(self, struct, unbound, case_sensitive=True):
        self.expr = Binder.bind(struct, unbound, case_sensitive)
        self.thread_local_data = threading.local()

    def _visitor(self):
        if not hasattr(self.thread_local_data, "visitors"):
            self.thread_local_data.visitors = Evaluator.EvalVisitor()

        return self.thread_local_data.visitors

    def eval(self, data):
        return self._visitor().eval(data, self.expr)

    class EvalVisitor(ExpressionVisitors.BoundExpressionVisitor):

        def __init__(self):
            super(Evaluator.EvalVisitor, self).__init__()
            self.struct = None

        def eval(self, row, expr):
            self.struct = row
            return ExpressionVisitors.visit(expr, self)

        def always_true(self):
            return True

        def always_false(self):
            return False

        def not_(self, result):
            return not result

        def and_(self, left_result, right_result):
            return left_result and right_result

        def or_(self, left_result, right_result):
            return left_result or right_result

        def is_null(self, ref):
            return ref.get(self.struct) is None

        def not_null(self, ref):
            return not (ref.get(self.struct) is None)

        def lt(self, ref, lit):
            return ref.get(self.struct) < lit.value

        def lt_eq(self, ref, lit):
            return ref.get(self.struct) <= lit.value

        def gt(self, ref, lit):
            return ref.get(self.struct) > lit.value

        def gt_eq(self, ref, lit):
            return ref.get(self.struct) >= lit.value

        def eq(self, ref, lit):
            return ref.get(self.struct) == lit.value

        def not_eq(self, ref, lit):
            return ref.get(self.struct) != lit.value

        def starts_with(self, ref, lit):
            return ref.get(self.struct).startswith(lit.value)

        def in_(self, ref, lit):
            raise NotImplementedError()

        def not_in(self, ref, lit):
            return not self.in_(ref, lit.value)
